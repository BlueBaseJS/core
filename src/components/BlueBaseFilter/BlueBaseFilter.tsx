import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { BlueBaseFilterProps } from '@bluebase/components';
import Loadable from 'react-loadable';
import React from 'react';
import { ReactLoadableLoading } from '../ReactLoadableLoading';

/**
 * # ðŸš‡ BlueBaseFilter
 *
 * Since filters in BlueBase are based on promises, it may be tedious to handle loading state,
 * error state, etc. It may also become a repetitive task.
 *
 * To solve this issue, we ship BlueBaseFilter component. Just pass name of filter, initial value,
 * and filter arguments as props. The final filtered value will be passed to the children function.
 * This component will handle loading and error states itself.
 *
 * ## Usage
 * ```jsx
 * <BlueBaseFilter filter="math" value={5} args={{ op: 'add' }} children={(val: number) => {
 * 	return <Text>{val}</Text>;
 * }} />
 * ```
 */
export class BlueBaseFilter extends React.PureComponent<BlueBaseFilterProps> {
	static contextType: React.Context<BlueBase> = BlueBaseContext;

	public static defaultProps = {
		args: {},
	};

	render() {
		const { filter, value, args, children } = this.props;

		const BB: BlueBase = this.context;

		const AsyncBlueBaseFilter = Loadable({
			loader: () => BB.Filters.run(filter, value, args),
			loading: ReactLoadableLoading,
			render(loadedValue: any) {
				return children(loadedValue);
			},
		});

		return <AsyncBlueBaseFilter />;
	}
}
