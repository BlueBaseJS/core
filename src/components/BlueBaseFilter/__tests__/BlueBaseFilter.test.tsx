import { BlueBaseApp } from '../../../';
import { BlueBaseFilter } from '../BlueBaseFilter';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Text } from 'react-native';

// beforeEach(() => {
// 	jest.resetModules();
// });

describe('BlueBaseFilter', () => {
	test(`should render a filtered value`, done => {
		const filters = {
			math: [
				{ key: 'add-fifteen', value: (val: number) => val + 15 },
				(val: number, { op }: { op: 'add' | 'subtract' } = { op: 'subtract' }) =>
					op === 'subtract' ? val - 5 : val + 5,
			],
		};

		const rendered: any = TestRenderer.create(
			<BlueBaseApp filters={filters}>
				<BlueBaseFilter filter="math" value={5} args={{ op: 'add' }}>
					{(val: number) => {
						return <Text>{val}</Text>;
					}}
				</BlueBaseFilter>
			</BlueBaseApp>
		);

		let text = rendered.root.findByType(Text);

		// expect(rendered.toJSON()).toMatchSnapshot();
		expect(text.instance.props.children).toBe('Loading');

		setTimeout(() => {
			// expect(rendered.toJSON()).toMatchSnapshot();

			text = rendered.root.findByType(Text);
			expect(text.instance.props.children).toBe(25);
			done();
		});
	});
});
