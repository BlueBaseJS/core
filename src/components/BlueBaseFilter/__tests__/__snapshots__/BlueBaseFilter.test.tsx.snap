// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlueBaseFilter should render a filtered value 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Loading
  </Text>
</View>
`;

exports[`BlueBaseFilter should render a filtered value 2`] = `
<Text>
  25
</Text>
`;
